DailyService

Drummss opened the pull request:
### Implementation of a DailyService
There is now a DailyService that will handle all things daily. Along with this, I added full localization for the daily command.

Commit:
Streaks Migration

- Daily streaks are now recorded in database rather than cache.

velddev commented on src/Miki/Modules/Accounts/AccountsModule.cs:
```diff
@@ -1033,30 +1033,29 @@ await e.ErrorEmbedResource("give_error_amount_unparsable")
             User u = await DatabaseHelpers.GetUserAsync(context, e.GetAuthor());
 
-            if(u == null)
+            if(u == null) // Check it the user exists.
```
What are you trying to explain with these comments that I cannot deduce from looking at the code?

velddev commented on src/Miki/Modules/Accounts/AccountsModule.cs:
```diff
@@ -1077,39 +1076,51 @@ await builder.ToEmbed()
 
-            int streak = 0;
+            /* Streak check. */
+            DailyStreak dailyStreak = await DailyStreak.GetAsync(context, u.Id);
```
Preferably we want to have no new raw Psql queries, please give a look at the UnitOfWork and Repository pattern. :-)

Drummss replied:
I think I'll need a bit of help with this but I have finished all of the other requested changes.

velddev commented on src/Miki/Modules/Accounts/AccountsModule.cs:
```diff
@@ -1077,39 +1076,51 @@ await builder.ToEmbed()
-            int streak = 0;
+            /* Streak check. */
+            DailyStreak dailyStreak = await DailyStreak.GetAsync(context, u.Id);
+            ICacheClient cache = e.GetService<ICacheClient>();
```
Please use `var` here

velddev commented on src/Miki/Modules/Accounts/AccountsModule.cs:
```diff
@@ -1077,39 +1076,51 @@ await builder.ToEmbed()
-            var cache = e.GetService<ICacheClient>();
+            if (await cache.ExistsAsync(redisKey)) // Check if the user had a cached streak and migrate to database.
+            {
+                dailyStreak.CurrentStreak = await cache.GetAsync<int>(redisKey);
```
Please use `GetAsync(...).ConfiguresAwait(false);`

velddev commented on src/Miki/Modules/Accounts/AccountsModule.cs:
```diff
@@ -1077,39 +1076,51 @@ await builder.ToEmbed()
             string redisKey = $"user:{e.GetAuthor().Id}:daily";
 
-            var cache = e.GetService<ICacheClient>();
+            if (await cache.ExistsAsync(redisKey)) // Check if the user had a cached streak and migrate to database.
```
Please use `ExistsAsync(...).ConfiguresAwait(false);`

velddev commented on src/Miki/Modules/Accounts/AccountsModule.cs:
```diff
@@ -1033,30 +1033,29 @@ await e.ErrorEmbedResource("give_error_amount_unparsable")
             }
 
+            /* Base variables */
             int dailyAmount = 100;
```
A good thing to do with these is to migrate them to constant values instead.

velddev commented on src/Miki/Modules/Accounts/AccountsModule.cs:
```diff
@@ -1033,30 +1033,29 @@ await e.ErrorEmbedResource("give_error_amount_unparsable")
-            var context = e.GetService<MikiDbContext>();
-
+            MikiDbContext context = e.GetService<MikiDbContext>();
             User u = await DatabaseHelpers.GetUserAsync(context, e.GetAuthor());
```
This is an old, deprecated way of getting users. The current prefered way is 

```cs
var userService = e.GetService<UserService>();
var user = userService.GetUserAsync(...);
```

velddev commented on src/Miki/Modules/Accounts/AccountsModule.cs:
```diff
@@ -1033,30 +1033,29 @@ await e.ErrorEmbedResource("give_error_amount_unparsable")
             int dailyStreakAmount = 20;
 
-            if(await u.IsDonatorAsync(context))
+            if (await u.IsDonatorAsync(context)) // Check donator status.
```
Please use `IsDonatorAsync(...).ConfiguresAwait(false);` And also please look at the donator module for a more up to date approach on checking if a user is a donator.

velddev commented on src/Miki/Modules/Accounts/AccountsModule.cs:
```diff
@@ -1077,39 +1076,51 @@ await builder.ToEmbed()
+            if(dailyStreak.CurrentStreak > 0)
             {
-                embed.AddInlineField("Streak!", $"You're on a {streak.ToFormattedString()} day daily streak!");
+                embed.AddInlineField("Streak!", $"You're on a {dailyStreak.CurrentStreak.ToFormattedString()} day daily streak!");
```
preferably use `{dailyStreak.CurrentStreak:N0}`.

velddev commented on src/Miki/Modules/DeveloperModule.cs:
```diff
@@ -339,5 +340,44 @@ public async Task BanUserAsync(IContext e)
+            [RequiresScope("developer")]
+            public async Task ResetDailyAsync(IContext e)
+            {
+                MikiDbContext context = e.GetService<MikiDbContext>();
```
If it is possible, avoid using raw database handling inside of the business context. I know it's still very much used everywhere, but there's an effort to no longer do this.

Commit:
UnitOfWork and Cleanup

Commit:
Merge branch 'miki-framework.3.0' into update_streaksmigration

Commit:
Add StreakService

velddev commented on src/Miki/Services/StreakService.cs:
```diff
@@ -0,0 +1,55 @@
+{
+    using System;
+    using System.Threading.Tasks;
+
```
What's with the spacing?

velddev commented on src/Miki/Services/StreakService.cs:
```diff
@@ -0,0 +1,55 @@
+
+    using Bot.Models;
+
+    using Framework;
```
Always use the full namespace using :-)

velddev commented on src/Miki/Services/StreakService.cs:
```diff
@@ -0,0 +1,55 @@
+
+    using Patterns.Repositories;
+
+    public class StreakService : IStreakService
```
There's a few issues with your service; It does not seem to elevate internal business logic to the service, but rather seem to work like a repository. I would recommend building a method that would take a user daily streak and increase it if valid. 

e.x.

```cs
public void IncreaseStreak(ulong userId) {
	var user = userService.GetUserById(userId);
	var streak = repository.Get(userId);
	if(IsValid(streak)) 
	{
		streak.value++;
 	}
	else 
	{
		streak = new DailyStreak();
	}
	repository.Edit(streak);

	transactionService.CreateTransaction(new TR { 
		Sender = 0L // Constant for the "miki bank"
		Receiver = (long)userId
		Amount = GetCurrentDailyAmount()
	});
	
	return streak;
}

velddev commented on src/Miki/Services/UserService.cs:
```diff
@@ -68,7 +78,9 @@ public interface IUserService : IDisposable
         ValueTask UpdateUserAsync(User user);
 
-        ValueTask<bool> UserIsBanned(long userId);
+        ValueTask<bool> UserIsBannedAsync(long userId);
```
Thanks for the naming refactor here +1

Drummss replied:
👍

velddev commented on src/Miki/Services/UserService.cs:
```diff
@@ -50,16 +49,27 @@ public async ValueTask<bool> UserIsBanned(long userId)
-        {
-            unitOfWork?.Dispose();
-        }
+            => unitOfWork?.Dispose();
```
Unit of work should never be null, avoid using `?` as it should throw an internal exception instead.

Drummss replied:
That's your code tbh.

velddev commented on src/Miki/Services/UserService.cs:
```diff
@@ -50,16 +49,27 @@ public async ValueTask<bool> UserIsBanned(long userId)
         {
-            return unitOfWork.CommitAsync();
+            var donatorRecord = await donatorRepository.GetAsync(userId);
+            if (donatorRecord == null)
```
replace with `return donatorRecord != null`

velddev commented on src/Miki/Modules/DeveloperModule.cs:
```diff
@@ -339,5 +343,48 @@ public async Task BanUserAsync(IContext e)
+        public class DailyEditCommand
+        {
+            [Command("resettimer")]
+            //[RequiresScope("developer")]
```
Don't commit commented code like this, instead make yourself a developer on your own database.

velddev commented on src/Miki/Modules/DeveloperModule.cs:
```diff
@@ -339,5 +343,48 @@ public async Task BanUserAsync(IContext e)
-	}
+
+        [Command("dailyedit")]
+        [RequiresScope("developer")]
```
Sadly this doesn't work yet 😔

velddev replied:
Please make a feature request in https://github.com/mikibot/miki.framework

velddev commented on src/Miki/Modules/Accounts/AccountsModule.cs:
```diff
@@ -1140,39 +1141,46 @@ await builder.ToEmbed()
-                    $"**{amount:N0}**", 
-                    $"`{u.Currency.ToFormattedString()}`"))
+                    $"**{finalAmount:N0}**", 
+                    $"`{user.Currency.ToFormattedString()}`"))
```
Replace `.ToFormattedString()` with `:N0`. Thanks! ❤️

velddev commented on src/Miki/Modules/Accounts/AccountsModule.cs:
```diff
@@ -1140,39 +1141,46 @@ await builder.ToEmbed()
             }
 
-            int amount = dailyAmount + (dailyStreakAmount * Math.Min(100, streak));
+            int finalAmount = (dailyAmount * donatorMultiplier) + ((dailyStreakAmount * donatorMultiplier) * Math.Min(100, (int)dailyStreak.CurrentStreak));
```
This seems to be a very long and complex line of code, might be useful to kind of move it to the service layer instead.

velddev commented on src/Miki/Modules/Accounts/AccountsModule.cs:
```diff
@@ -1140,39 +1141,46 @@ await builder.ToEmbed()
+            {
+                dailyStreak.CurrentStreak++;
+                dailyStreak.LastStreakTime = DateTime.UtcNow;
+            } else
```
`else` should be on a newline

velddev commented on src/Miki/Modules/Accounts/AccountsModule.cs:
```diff
@@ -1140,39 +1141,46 @@ await builder.ToEmbed()
+            string redisKey = $"user:{user.Id}:daily";
 
-            var cache = e.GetService<ICacheClient>();
+            if (await cache.ExistsAsync(redisKey).ConfigureAwait(false))
```
Avoid using cache operations directly in commands. Rather, let the service handle it

velddev commented on src/Miki/MikiBotApp.cs:
```diff
@@ -6,12 +6,17 @@
     using System.Reflection;
     using System.Resources;
     using System.Threading.Tasks;
+
```
Please keep the current style consistent with the already defined style

velddev commented on src/Miki/Miki.csproj:
```diff
@@ -43,6 +43,7 @@
     <PackageReference Include="NCalc.NetCore" Version="1.0.1" />
     <PackageReference Include="Newtonsoft.Json" Version="12.0.2" />
     <PackageReference Include="Npgsql" Version="4.1.2" />
+    <PackageReference Include="Npgsql.EntityFrameworkCore.PostgreSQL" Version="3.0.1" />
```
Why was this required here?

velddev reviewed (changes requested):
It seems to be improving a lot, however; we're not there yet. 👍 great progress man!

Commit:
Staging for catchup with master

- Started work on implementing a DailyService to handle all things daily.

Commit:
Merge branch 'master' into update_streaksmigration

Commit:
Fixed merge errors

- Fix: AccountsModule:GetDailyAsync code updated to work with transaction service addition.
- Fix: Merge duplications.

Commit:
Merge branch 'master' into update_streaksmigration

Commit:
Daily command working with database

- Add: DailyService now handles everything daily.
- Add: Daily properties to AppProps.

- Remove: dailyedit.resetstreak command.

- Rename: dailyedit.resettimer > dailyedit.reset

Commit:
Unit tests for DailyService

- Add: DailyService tests.

Commit:
Moved transaction into DailyService.

Commit:
Added some words to solution dictionary.

Words added were "Miki", "Mekos" and "Patreon".

Commit:
Added full localization for the daily command.

Commit:
Merge branch 'master' into update_streaksmigration

Commit:
Remove outdated todo... whoops.

Commit:
Repaired daily streak reset bug

- Add: Actually added the code that would make the daily streak reset after 24 hours... and I added a unit test for it.

velddev commented on src/Miki/Languages/eng.resx:
```diff
@@ -2109,7 +2109,7 @@ To give someone reputation, do `&gt;rep &lt;mention&gt; ["all"|amount]`</value>
     <value>You can not use an amount less than zero!</value>
   </data>
-  <data name="daily_received" xml:space="preserve">
+  <data name="miki_module_accounts_daily_received" xml:space="preserve">
```
This is an old method of handling language strings; go for the shorter version instead.

velddev commented on src/Miki/Languages/eng.resx:
```diff
@@ -2308,4 +2308,23 @@ To read more about the miki scripting language. Check out https://github.com/mik
   <data name="error_pasta_null" xml:space="preserve">
     <value>Pasta was not found.</value>
   </data>
+  <data name="miki_module_accounts_daily_appreciate_list" xml:space="preserve">
```
Same for these

velddev replied:
Also; this one is not only used in daily.

velddev commented on src/Miki/Modules/Accounts/AccountsModule.cs:
```diff
@@ -1136,88 +1138,43 @@ private void ThrowOnUserNull(IDiscordGuildUser user)
-
             var embed = new EmbedBuilder()
-                .SetTitle("💰 Daily")
+                .SetTitle(e.GetLocale().GetString("miki_module_accounts_daily_title"))
```
replace to `daily_claimed_title`

velddev commented on src/Miki/Modules/Accounts/AccountsModule.cs:
```diff
@@ -1136,88 +1138,43 @@ private void ThrowOnUserNull(IDiscordGuildUser user)
-                    "daily_received", 
-                    $"**{amount:N0}**", 
-                    $"`{(u.Currency):N0}`"))
+                    "miki_module_accounts_daily_received", 
```
replace to `daily_claimed`

velddev commented on src/Miki/Languages/eng.resx:
```diff
@@ -2308,4 +2308,23 @@ To read more about the miki scripting language. Check out https://github.com/mik
+    <value>Streak!</value>
+  </data>
+  <data name="miki_module_accounts_daily_title" xml:space="preserve">
+    <value>💰 Daily</value>
```
Avoid using emoji in language files

velddev commented on src/Miki/Languages/eng.resx:
```diff
@@ -2308,4 +2308,23 @@ To read more about the miki scripting language. Check out https://github.com/mik
+    <value>Appreciate Miki?</value>
+  </data>
+  <data name="miki_module_accounts_daily_claimed" xml:space="preserve">
+    <value>You already claimed your daily today! Please wait another `{0}` before using it again.</value>
```
Avoid using markdown styling in language files

velddev commented on src/Miki/Services/DailyService.cs:
```diff
@@ -0,0 +1,159 @@
+﻿using Miki.Cache;
+using Miki.Core.Migrations;
+using Miki.Logging;
```
Move these inside the namespace

velddev commented on src/Miki/Services/DailyService.cs:
```diff
@@ -0,0 +1,159 @@
+            => unitOfWork.Dispose();
+    }
+
+    public class DailyClaimResponse
```
Move public classes in its own file

Commit:
Codacy compliancy and appreciation fix

- Changed code to be more compliant with Codacy.
- Updated the random daily appreciation to scale with the amount of members in the list.

Commit:
Merge branch 'master' into update_streaksmigration

Commit:
Merge branch 'master' into update_streaksmigration

Commit:
Update to new Localization and Cleanup

- Change: Moved DailyClaimResponse into it's own file.
- Change: DailyService and DailyClaimResponse are in their own namespace now.
- Add: Daily localization strings into eng.json.

Commit:
Added test for checking daily cache migration

Commit:
Character limit for roll command

Commit:
Reworked cached streak migration test

- Removed duplicated namespace reference in MikiBotApp.cs
- Cleaned code to be compliant with Codacy

Commit:
Remove old language files
